example/src/main/java/io/netty/example/tunechoserver/InetUtils.java

```diff
@@ -0,0 +1,329 @@
+            if (ipHeader.headerLength < IPV4_MIN_HEADER_LENGTH) {
+                throw new IllegalArgumentException("Invalid IPv4 header length");
+            }
+
+            // Fail if the packet is too small to contain the IPv4 header.
+            if (packetBuf.readableBytes() < ipHeaderOffset + ipHeader.headerLength) {
+                throw new IllegalArgumentException("IP packet too small");
+            }
+
+            // Get the total length of the packet.
+            int totalLength = packetBuf.getUnsignedShort(ipHeaderOffset + IPV4_HEADER_OFFSET_TOTAL_LENGTH);
+
+            // Fail if the packet is smaller than the total length.
+            if (packetBuf.readableBytes() < ipHeaderOffset + totalLength) {
+                throw new IllegalArgumentException("IP packet too small");
+            }
+
+            // Compute the payload length
+            ipHeader.payloadLength = totalLength - ipHeader.headerLength;
+
+            // Get the protocol.
+            ipHeader.protocol = packetBuf.getUnsignedByte(ipHeaderOffset + IPV4_HEADER_OFFSET_PROTOCOL);
+
+            // Get the source and destination addresses.
+            ipHeader.sourceAddress = new byte[4];
+            ipHeader.destAddress = new byte[4];
+            packetBuf.getBytes(ipHeaderOffset + InetUtils.IPV4_HEADER_OFFSET_SOURCE_ADDR, ipHeader.sourceAddress);
+            packetBuf.getBytes(ipHeaderOffset + InetUtils.IPV4_HEADER_OFFSET_DEST_ADDR, ipHeader.destAddress);
+
+            // Get IP checksum.
+            ipHeader.checksum = packetBuf.getUnsignedShort(ipHeaderOffset + IPV4_HEADER_OFFSET_CHECKSUM);
+
+        // Otherwise the packet type is unknown.
+        } else {
+            throw new UnsupportedOperationException(
+                    "IP version unsupported: " + ipVersionToString(ipHeader.ipVersion));
+        }
+
+        return ipHeader;
+    }
+
+    public static final class ICMPHeader {
+        public int type;
+        public int code;
+        public int checksum;
+    }
+
+    public static ICMPHeader decodeICMPHeader(ByteBuf packetBuf) {
+        return decodeICMPHeader(packetBuf, packetBuf.readerIndex());
+    }
+
+    public static ICMPHeader decodeICMPHeader(ByteBuf packetBuf, int headerOffset) {
+        // Fail if the packet is too small.
+        if (packetBuf.readableBytes() < headerOffset + ICMP_HEADER_LENGTH) {
+            throw new IllegalArgumentException("ICMP packet too small");
+        }
+
+        ICMPHeader icmpHeader = new InetUtils.ICMPHeader();
+        icmpHeader.type = packetBuf.getUnsignedByte(headerOffset + ICMP_HEADER_OFFSET_TYPE);
+        icmpHeader.code = packetBuf.getUnsignedByte(headerOffset + ICMP_HEADER_OFFSET_CODE);
+        icmpHeader.checksum = packetBuf.getUnsignedShort(headerOffset + ICMP_HEADER_OFFSET_CHECKSUM);
+
+        return icmpHeader;
+    }
+
+    public static final class UDPHeader {
+        public int sourcePort;
+        public int destPort;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make destPort a static final constant or non-public and provide accessors if needed. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)